### DISCUSSION TITLE
How to handle non-listening devices that claim not to support WakeUp CC

### DISCUSSION THREAD

**Discussion by AlCalzone** [zwave-js/zwave-js#1537] [General] [answered]
@hanskroner #1536 is an interesting case. The device is non-listening, does not support Z-Wave+, and does not list the WakeUp CC in its NIF:
```
17:44:12.657 CNTRLR   finished adding node 8:
                        basic device class:    Controller
                        generic device class:  Remote Controller
                        specific device class: Portable Remote Controller
                        supported CCs: 
                        · Version (0x86)
                        · Manufacturer Specific (0x72)
                        · Configuration (0x70)
                        · Association Command Configuration (0x9b)
                        controlled CCs: 
                        · Basic (0x20)
                        · Association (0x85)
                        · Scene Activation (0x2b)
                        · Multilevel Switch (0x26)

...

17:44:12.694 DRIVER « [RES] [GetNodeProtocolInfo]
                        payload: 0x129600010101
17:44:12.699 CNTRLR « [Node 008] received response for protocol info:
                      basic device class:    Controller
                      generic device class:  Remote Controller
                      specific device class: Portable Remote Controller
                      is a listening device: false
                      is frequent listening: false
                      is a routing device:   false
                      is a secure device:    unknown
                      is a beaming device:   true
                      maximum baud rate:     40000 kbps
                      version:               3
```

By default, zwave-js assumes that a non-listening device supports the WakeUp CC, even if it is not in the NIF. However, this one responds to the CC version query with 0 - causing zwave-js to revert its previous assumption. It does send Wakeup Notifications though.

I'm not sure what's the best strategy here.

**Comment by hanskroner** [comment, marked as answer]
> @hanskroner #1536 is an interesting case. The device is non-listening, does not support Z-Wave+, and does not list the WakeUp CC in its NIF:

I would guess the minimote was intended to be a stand-alone controller back when it was manufactured. It doesn't seem like the manufacturer intended for this device to be communicated with, unless the user manually put it into "extended awake mode" so another device could reach it

> By default, zwave-js assumes that a non-listening device supports the WakeUp CC, even if it is not in the NIF.

Certification does not allow for classes that are not listed as supported to be actually supported - I don't think this assumption is a good idea.

> However, this one responds to the CC version query with 0 - causing zwave-js to revert its previous assumption. It does send Wakeup Notifications though.

It sends Wake-Up Notifications to whom? By not advertising support for the Wake-Up Command Class, an including controller would not have set the Wake-Up Interval or Destination. Is it maybe sending it out as a `BROADCAST` in some sort of "Hail Mary" attempt? `BROADCAST`s aren't routed or repeated in Z-Wave, so don't count on receiving these messages unless the minimote is in direct range of the controller.

> I'm not sure what's the best strategy here.

The device is describing itself to the controller saying there's no way to reach it periodically, so I would suggest not trying to reach it periodically. The intention with the device seems to be that a user will put it in "extended awake" mode by pressing some button combination whenever they want to send something to it. This should be doable if _zwave-js_'s transmission strategy is not trying to queue up messages to this device in a wake-up queue (which it shouldn't be doing, since the device didn't advertise support for the Wake-Up Command Class).

**Reply by AlCalzone** [reply]
> It sends Wake-Up Notifications to whom? 

Like you expect - as broadcasts.

> The device is describing itself to the controller saying there's no way to reach it periodically, so I would suggest not trying to reach it periodically.

For the sake of argument (the device actually behaves like a normal sleeping device if we assume support for the WakeUp CC): Any command we send (whether it is library-initiated or user-initiated) to a device that is not listening and not supporting Wakeup will just time out, except if the device is coincidentally listening (or in such an extended awake mode). Is that the intended behavior?

> This should be doable if zwave-js's transmission strategy is not trying to queue up messages to this device in a wake-up queue (which it shouldn't be doing, since the device didn't advertise support for the Wake-Up Command Class).

Sounds reasonable, but that requires me to rework the interview strategy a bit (we've already discussed that), since we'd end up with a bunch of queries on startup that will all time out.

**Reply by hanskroner** [reply]
> For the sake of argument (the device actually behaves like a normal sleeping device if we assume support for the WakeUp CC): Any command we send (whether it is library-initiated or user-initiated) to a device that is not listening and not supporting Wakeup will just time out, except if the device is coincidentally listening (or in such an extended awake mode). Is that the intended behavior?

In my opinion, yes.

> Sounds reasonable, but that requires me to rework the interview strategy a bit (we've already discussed that), since we'd end up with a bunch of queries on startup that will all time out.

If by "startup" I'm understanding _zwave-js_'s startup, then yes - it sounds like there'll be a few queries that hang around until timing out. I know I'm beating a dead horse here, but because clients need to deal with the fact that a Z-Wave network could be made up of entirely wake-up-based devices that you can't refresh on startup, this "interview everything after a restart" isn't an approach I recommend taking - I favor reporting the device's states as "unknown" and letting clients ask you for the information they're interested in, when they need it (and "I still don't know" should be a perfectly valid answer to those questions).

**Reply by AlCalzone** [reply]
> I know I'm beating a dead horse here, but because clients need to deal with the fact that a Z-Wave network could be made up of entirely wake-up-based devices that you can't refresh on startup, this "interview everything after a restart" isn't an approach I recommend taking

I'm working on that, but there's baby steps I have to take without breaking everything.
